File: public/python/combined.py
from pipes_utils import PipeType
from constraints.no_half_connections import (
    validator_h as no_half_connections_validator_h,
    validator_v as no_half_connections_validator_v,
    pruner_h as no_half_connections_pruner_h,
    pruner_v as no_half_connections_pruner_v,
)
from constraints.no_cycles import validator as tree_validator, pruner as tree_pruner
from constraints.connected import (
    validator as connected_validator,
    pruner as connected_pruner,
)
from csp import CSP, Variable, Constraint

# Creates CSP object
# implementation details can be found in the combined.py section in the report


def generate_domain(top: bool, right: bool, bottom: bool, left: bool) -> list[PipeType]:
    """
    Generate a domain based on the four boolean flags:
    if i == 0: 0 is false (top)
    if i == n-1: 2 is false (bottom)

    if j == 0: 3 is false (left)
    if j == n-1: 1 is false (right)

    :param top: A boolean indicating if the top of the pipe is blocked.
    :param right: A boolean indicating if the right of the pipe is blocked.
    :param bottom: A boolean indicating if the bottom of the pipe is blocked.
    :param left: A boolean indicating if the left of the pipe is blocked.
    :return: A list of PipeType objects representing the domain.

    """
    # all the possible domains for pipes in the pipes puzzle
    # (True, True, True, True) and (False, False, False, False) are omitted - they represent all connections or no connections, which are immune to rotations.
    domain: list[PipeType] = [
        (True, True, True, False),
        (True, True, False, True),
        (True, True, False, False),
        (True, False, True, True),
        (True, False, True, False),
        (True, False, False, True),
        (True, False, False, False),
        (False, True, True, True),
        (False, True, True, False),
        (False, True, False, True),
        (False, True, False, False),
        (False, False, True, True),
        (False, False, True, False),
        (False, False, False, True),
    ]
    if top:
        # remove pipes that point up if the pipe is at the top of the grid
        domain = [pipe for pipe in domain if not pipe[0]]
    if bottom:
        # remove pipes that point down if the pipe is at the bottom of the grid
        domain = [pipe for pipe in domain if not pipe[2]]
    if right:
        # remove pipes that point right if the pipe is on the right side of the grid
        domain = [pipe for pipe in domain if not pipe[1]]
    if left:
        # remove pipes that point left if the pipe is on the left side of the grid
        domain = [pipe for pipe in domain if not pipe[3]]
    return domain


def create_pipes_csp(
    n: int,
) -> CSP:
    variables: list[Variable] = []

    # initialize variable objects
    for i in range(n):
        row: list[Variable] = []
        for j in range(n):
            top = i == 0
            bottom = i == n - 1
            left = j == 0
            right = j == n - 1
            var = Variable(
                location=i * n + j,
                domain=generate_domain(top, right, bottom, left),
            )
            row.append(var)
        variables += row

    all_cons: list[Constraint] = []

    # create binary constraints for no blocking
    no_half_connections_cons: list[Constraint] = []
    # start with horizontal cons
    no_half_connections_h: list[Constraint] = []
    for i in range(n):
        for j in range(n - 1):
            left = variables[i * n + j]
            right = variables[i * n + j + 1]
            scope = [left, right]
            name = f"no half-connections horizontal {i * n + j, i * n + j + 1}"
            no_half_connections_h.append(
                Constraint(
                    name,
                    no_half_connections_validator_h,
                    no_half_connections_pruner_h,
                    scope,
                )
            )

    # vertical cons
    no_half_connections_v: list[Constraint] = []
    for i in range(n - 1):
        for j in range(n):
            above = variables[i * n + j]
            below = variables[(i + 1) * n + j]
            scope = [above, below]
            name = f"no half-connections vertical {i * n + j, (i + 1) * n + j}"
            no_half_connections_v.append(
                Constraint(
                    name,
                    no_half_connections_validator_v,
                    no_half_connections_pruner_v,
                    scope,
                )
            )

    # add cons
    no_half_connections_cons += no_half_connections_h + no_half_connections_v

    # create tree constraint
    tree_con: Constraint = Constraint("tree", tree_validator, tree_pruner, variables)

    connected_con: Constraint = Constraint(
        "connected", connected_validator, connected_pruner, variables
    )
    all_cons = no_half_connections_cons + [tree_con, connected_con]

    return CSP(f"Pipes_{n}x{n}", variables, all_cons)

File: utils/csp/combined.ts
import { PipeType } from "./utils";
import {
  noHalfConnectionsValidatorH,
  noHalfConnectionsValidatorV,
  noHalfConnectionsPrunerH,
  noHalfConnectionsPrunerV,
  noCyclesValidator,
  noCyclesPruner,
  connectedValidator,
  connectedPruner,
} from "./constraints";
import { CSP, Variable, Constraint } from "./csp";

/**
 * Generate a domain based on the four boolean flags:
 * if i == 0: 0 is false (top)
 * if i == n-1: 2 is false (bottom)
 * if j == 0: 3 is false (left)
 * if j == n-1: 1 is false (right)
 *
 * @param top A boolean indicating if the top of the pipe is blocked.
 * @param right A boolean indicating if the right of the pipe is blocked.
 * @param bottom A boolean indicating if the bottom of the pipe is blocked.
 * @param left A boolean indicating if the left of the pipe is blocked.
 * @returns A list of PipeType objects representing the domain.
 */
export function generateDomain(
  top: boolean,
  right: boolean,
  bottom: boolean,
  left: boolean
): PipeType[] {
  // all the possible domains for pipes in the pipes puzzle
  // [true, true, true, true] and [false, false, false, false] are omitted - they represent all connections or no connections, which are immune to rotations.
  let domain: PipeType[] = [
    [true, true, true, false],
    [true, true, false, true],
    [true, true, false, false],
    [true, false, true, true],
    [true, false, true, false],
    [true, false, false, true],
    [true, false, false, false],
    [false, true, true, true],
    [false, true, true, false],
    [false, true, false, true],
    [false, true, false, false],
    [false, false, true, true],
    [false, false, true, false],
    [false, false, false, true],
  ];

  if (top) {
    // remove pipes that point up if the pipe is at the top of the grid
    domain = domain.filter((pipe) => !pipe[0]);
  }
  if (bottom) {
    // remove pipes that point down if the pipe is at the bottom of the grid
    domain = domain.filter((pipe) => !pipe[2]);
  }
  if (right) {
    // remove pipes that point right if the pipe is on the right side of the grid
    domain = domain.filter((pipe) => !pipe[1]);
  }
  if (left) {
    // remove pipes that point left if the pipe is on the left side of the grid
    domain = domain.filter((pipe) => !pipe[3]);
  }

  return domain;
}

/**
 * Creates a CSP for the pipes puzzle
 * @param n The size of the grid (n x n)
 * @returns A CSP object representing the pipes puzzle
 */
export function createPipesCSP(n: number): CSP {
  const variables: Variable[] = [];

  // initialize variable objects
  for (let i = 0; i < n; i++) {
    const row: Variable[] = [];
    for (let j = 0; j < n; j++) {
      const top = i === 0;
      const bottom = i === n - 1;
      const left = j === 0;
      const right = j === n - 1;

      const v = new Variable(
        i * n + j,
        generateDomain(top, right, bottom, left)
      );

      row.push(v);
    }
    variables.push(...row);
  }

  const allConstraints: Constraint[] = [];

  // create binary constraints for no blocking
  const noHalfConnectionsConstraints: Constraint[] = [];

  // start with horizontal constraints
  const noHalfConnectionsH: Constraint[] = [];
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n - 1; j++) {
      const left = variables[i * n + j];
      const right = variables[i * n + j + 1];
      const scope = [left, right];
      const name = `no half-connections horizontal ${i * n + j}, ${
        i * n + j + 1
      }`;

      noHalfConnectionsH.push(
        new Constraint(
          name,
          noHalfConnectionsValidatorH,
          noHalfConnectionsPrunerH,
          scope
        )
      );
    }
  }

  // vertical constraints
  const noHalfConnectionsV: Constraint[] = [];
  for (let i = 0; i < n - 1; i++) {
    for (let j = 0; j < n; j++) {
      const above = variables[i * n + j];
      const below = variables[(i + 1) * n + j];
      const scope = [above, below];
      const name = `no half-connections vertical ${i * n + j}, ${
        (i + 1) * n + j
      }`;

      noHalfConnectionsV.push(
        new Constraint(
          name,
          noHalfConnectionsValidatorV,
          noHalfConnectionsPrunerV,
          scope
        )
      );
    }
  }

  // add constraints
  noHalfConnectionsConstraints.push(
    ...noHalfConnectionsH,
    ...noHalfConnectionsV
  );

  // create tree constraint
  const treeConstraint = new Constraint(
    "tree",
    noCyclesValidator,
    noCyclesPruner,
    variables
  );

  // create connected constraint
  const connectedConstraint = new Constraint(
    "connected",
    connectedValidator,
    connectedPruner,
    variables
  );

  // add all constraints
  allConstraints.push(
    ...noHalfConnectionsConstraints,
    treeConstraint,
    connectedConstraint
  );

  return new CSP(`Pipes_${n}x${n}`, variables, allConstraints);
}
